Style-Guide/Code-Layout-and-Formatting.md

```diff
@@ -166,30 +197,36 @@ $variable=Get-Content $FilePath -Wai:($ReadCount-gt0) -First($ReadCount*5)
 $variable = Get-Content -Path $FilePath -Wait:($ReadCount -gt 0) -TotalCount ($ReadCount * 5)
 ```
 
-#### Spaces around special characters
+### Spaces around special characters
 
-White-space is (mostly) irrelevant to PowerShell, but its proper use is the key to writing easily readable code.
+White-space is (mostly) irrelevant to PowerShell, but its proper use is key to writing easily readable code.
 
 Use a single space after commas and semicolons, and around pairs of curly braces.
 
-Avoid extra spaces inside parenthesis or square braces.
+Avoid unnecessary extra spaces inside parenthesis or square braces.
```

Comment:
Weakens the statement

Reply:
Hm, perhaps. The original statement seems to contradict prior assertions, however, implying vaguely that all space that *can* be omitted inside such expressions *should*... Which i do not think is wise or consistent with prior assertions.

Reply:
Well, frankly: you should **not** put spaces _inside_ parentheses or square braces -- although it's sometimes acceptable when you have multiple parentheses together (particularly when it wraps across multiple lines).

I mean, for example, don't write: 
```posh
foreach ( $x in Get-Content $file ) {
    if ( $Variable[ $x ] ) {
```
Instead write:  
```posh
foreach ($x in Get-Content $file) {
    if ($Variable[$x]) {
```

What do you think of changing this to read:

> Sub-expressions and scriptblocks, on the other hand, should usually have a single space on the inside of the braces or parentheses to make them stand out and improve readability. Note that of course, sub-expressions and variable delimiters _inside strings_ should not include additional space, because it affects the actual string.
> 
> ```posh
> foreach ($x in $( Get-Content $file )) {
>     Write-Host "The number [Math]::Pow(2, ${x}) = $( [Math]::Pow(2, $x) )"
> ```

Although after some performance testing, we've basically stopped writing `${x}` unless we're using non-standard variable names, and are using `$($x)` when it's necessary to separate the variable from the text around it. 

In fact, at work we've decided **all** variables in strings should be wrapped in a sub-expression, so that we can more easily refactor using search-and-replace in our editor without worrying about whether there's an instance in a string when we change `$Thing` to `$Thing.Name` or whatever.

Reply:
Well, for one the `foreach ($x in $( Get-Content $file )) {` is no different to `foreach ($x in ( Get-Content $file )) {` in general (in fact, most people I've spoken to have no idea what `$()` really adds outside of the standard usage in strings).

But overall I tend to agree, mostly. Excessive spacing in an expression can sometimes *detract*, though, as you mention, so there's a bit of a "do what works best for you" element here as well for some of the more complex expressions.

Example:

```powershell
for (int $i = ($j - 1) ; ($i -gt 19) -or ($i -lt 1) ; ($i + ($j / 2)) { }
```
(Now granted, that *is* a highly contrived example, but the point is mainly that there are a great many places you could opt to place additional spaces -- but adding spaces *everywhere* makes the expression **more** difficult to read, because the expressions lose their grouped appearance which lends some intuitive clarity ("things that *look* like they're together work together" sort of idea).

Reply:
We actually standardized on `$($x)` over `${x}` for two reasons:
1. It appears that the sub-expression is actually faster (although this is hard to measure)
2. It kept coming up that someone would try to refactor variables into objects and we'd end up trying to do a refactor of, say `$userName` to `$User.Name` -- you can rename $userName to $User.Name with a simple search and replace, but only if you *never* wrote ${userName} anywhere...

Not trying to put that in the best practices book (yet) though 😉